### PR TITLE
fix: openid ignores missing redirect uri

### LINKED ISSUE #685: Investigate missing `redirect_uri` check when performing OIDC flows
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/fosite/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/fosite/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Cloud](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

[Open ID Connect Core specification](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) says:

```
redirect_uri
REQUIRED. Redirection URI to which the response will be sent. This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider, with the matching performed as described in Section 6.2.1 of [[RFC3986]](https://openid.net/specs/openid-connect-core-1_0.html#RFC3986) (Simple String Comparison). When using this flow, the Redirection URI SHOULD use the https scheme; however, it MAY use the http scheme, provided that the Client Type is confidential, as defined in Section 2.1 of OAuth 2.0, and provided the OP allows the use of http Redirection URIs in this case. The Redirection URI MAY use an alternate scheme, such as one that is intended to identify a callback into a native application.
```

It does however not appear as if we are checking for this condition, as OAuth2 has the redirect URI marked as optional.

### Reproducing the bug

This should be first confirmed with an integration test case

### Relevant log output

_No response_

### Relevant configuration

_No response_

### Version

-

### On which operating system are you observing this issue?

_No response_

### In which environment are you deploying?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by james-d-elliott**
Fixes an issue where Authorize Requests which were intended for an OpenID Connect 1.0 client would incorrectly be allowed when missing the redirect URI when it's required by the specification.

## Related Issue or Design Document

Fixes #685

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [x] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation within the code base (if appropriate).

## Further comments

The added tests fail prior to the fix and pass after.

### CODE REVIEW

File: authorize_request_handler.go [hunk 1]
```diff
@@ -163,6 +163,17 @@ func (f *Fosite) validateAuthorizeRedirectURI(_ *http.Request, request *Authoriz
...
 	rawRedirURI := request.Form.Get("redirect_uri")
 
+	// This ensures that the 'redirect_uri' parameter is present for OpenID Connect 1.0 authorization requests as per:
+	//
+	// Authorization Code Flow - https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+	// Implicit Flow - https://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthRequest
+	// Hybrid Flow - https://openid.net/specs/openid-connect-core-1_0.html#HybridAuthRequest
+	//
+	// Note: as per the Hybrid Flow documentation the Hybrid Flow has the same requirements as the Authorization Code Flow.
+	if len(rawRedirURI) == 0 && request.GetRequestedScopes().Has("openid") {
```
Comment (aeneasr) [inline authorize_request_handler.go:173]: Is this not already handled by the openid handler itself?

Comment (james-d-elliott) [reply authorize_request_handler.go:173]: I have added an integration test which shows this is unfortunately not the case. It only returns an error if the client has multiple redirect URI's. The issue is that OAuth 2.0 I believe allows assuming the Redirect URI if absent provided there is only one, however OpenID Connect 1.0 explicitly requires the redirect URI parameter and has no such leeway.

It should also be noted that OAuth 2.1 does not alter this behavior. The only alteration by my reading is in OpenID Connect 1.0.

Comment (james-d-elliott) [reply authorize_request_handler.go:173]: Also see the linked issue, you were actually the person to raise the issue. Maybe that's helpful?

**Review by mitar — APPROVED** [review]

File: authorize_request_handler.go (hunk 1 above)
Comment (mitar) [reply authorize_request_handler.go:175]: I checked this and I agree with @james-d-elliott: OAuth 2.0 says that if you included Redirect URI in authentication request, then it has also be included in the token request. And Fosite currently checks that.

But OIDC spec says that Redirect URI is required in the authentication request. And this PR adds this check.

File: authorize_request_handler.go [hunk 2]
```diff
@@ -363,27 +378,31 @@ func (f *Fosite) newAuthorizeRequest(ctx context.Context, r *http.Request, isPAR
...
 		return request, errorsx.WithStack(ErrRegistrationNotSupported)
 	}
 
-	if err := f.validateResponseTypes(r, request); err != nil {
+	if err = f.validateResponseTypes(r, request); err != nil {
 		return request, err
 	}
 
-	if err := f.validateResponseMode(r, request); err != nil {
+	if err = f.validateResponseMode(r, request); err != nil {
```
Comment (mitar) [inline authorize_request_handler.go:405]: Why these changes? I think they are unrelated and just change the style?

File: authorize_request_handler.go [hunk 3]
```diff
@@ -363,27 +378,31 @@ func (f *Fosite) newAuthorizeRequest(ctx context.Context, r *http.Request, isPAR
 		return request, err
 	}
 
-	if err := f.validateAuthorizeRedirectURI(r, request); err != nil {
+	if err = f.parseAuthorizeScope(r, request); err != nil {
```
Comment (mitar) [inline authorize_request_handler.go:381]: I think that instead of splitting `validateAuthorizeScope` into `parseAuthorizeScope` and `validateAuthorizeScope`, it would be easier if you swapped the order of validation, first call (original) `validateAuthorizeScope` and then (updated) `validateAuthorizeRedirectURI`. In a way, if `validateAuthorizeRedirectURI` depends on parsed scopes, it is better if `validateAuthorizeRedirectURI` depends on parsed and validated scopes.

**Review by mitar — CHANGES_REQUESTED** [review]

File: authorize_request_handler.go (hunk 1 above)
Comment (aeneasr) [reply authorize_request_handler.go:175]: Thank you both - I checked and indeed the expectation is that the redirect-url is properly validated before we call `HandleAuthorizeEndpointRequest`. Given that this is openID specific though, this should be moved to the openid handler instead. I'll do that now

Comment (aeneasr) [reply authorize_request_handler.go:175]: I'm creating a separate PR as this is a breaking change for existing clients and it requires dedicated care for merging that to production on our infrastructure!

Comment (aeneasr) [reply authorize_request_handler.go:175]: As far as I can tell, this implementation does not respect hybrid or implicit flows of OIDC, so it definitely makes sense to move that to the appropriate handlers in my view.

**Comment by aeneasr** [comment]
Closing in favor of https://github.com/ory/fosite/pull/814

File: authorize_request_handler.go (hunk 1 above)
Comment (james-d-elliott) [reply authorize_request_handler.go:175]: That sounds like a solid plan, glad this will get merged regardless.
